### PR TITLE
Fixes possible deadlock upon epoch close

### PR DESCRIPTION
**Pull request by mustermeiszer**
The following PR closes https://github.com/centrifuge/centrifuge-chain-internal/issues/11

NOTE: Most of the changes are tests ;-)

### CODE REVIEW

**Comment by mustermeiszer** [comment]
There is an exiting issue currently in the codebase with how we handle bounds of pool parameters. I did not wanted to include this here. This issue here https://github.com/centrifuge/centrifuge-chain/issues/718 describes the needed changes and we should address this before altair launch.

**Comment by offerijns** [comment]
I'm not really sure I follow why we need the min submission time. Why can't we just immediately accept a no execution solution when the epoch goes into a submission period? This is how it also worked with Tinlake on Ethereum: https://github.com/centrifuge/tinlake/blob/develop/src/lender/test/coordinator-submitEpoch.sol#L250

**Comment by mustermeiszer** [comment]
@offerijns WDYT? Am I missing something. We could also allow to always check for solutions to be `new_solution >= prev_solution`. This would mean, that one can always start the `challenge_period` by submitting a zero-fulfillment-solution. 

My feeling is, that this should be okayish, as the only attack vector would be to prevent/make-it-harder-for fulfillments.

**Comment by offerijns** [comment]
We could also allow to always check for solutions to be `new_solution >= prev_solution`. This would mean, that one can always start the `challenge_period` by submitting a zero-fulfillment-solution.
> 
> My feeling is, that this should be okayish, as the only attack vector would be to prevent/make-it-harder-for fulfillments.

I'll try to take another more detailed look at this tomorrow, but I think this is indeed how it worked with Tinlake on Ethereum, because of the `>=` here: https://github.com/centrifuge/tinlake/blob/develop/src/lender/coordinator.sol#L416

**Comment by branan** [comment]
I think the `>=` implementation is cleaner, but I don't have _super_ strong feelings

**Comment by mustermeiszer** [comment]
I would be fine with both. We just must agree upon that we allow zero-solutions to trigger the challenge time. Wdyt? Can we agree on this here? Then I can close this next week.

**Comment by offerijns** [comment]
> I would be fine with both. We just must agree upon that we allow zero-solutions to trigger the challenge time. Wdyt? Can 
we agree on this here? Then I can close this next week.

Thought about this some more, I think allowing zero solutions to solve this deadlock makes sense. The goal of the challenge period is to prevent suboptimal solutions from being accepted, by giving interested actors a reasonable chance to submit a better solution. As long as the challenge period is sufficiently long, then a zero-solution does not change this, anyone will have time to submit a better solution.

**Comment by mustermeiszer** [comment]
@branan, @offerijns I adapted the PR to the above. If you give green lights for this I will merge.

File: pallets/pools/src/lib.rs [hunk 1]
```diff
@@ -1027,7 +1027,7 @@ pub mod pallet {
 				let new_solution = Self::score_solution(&pool, &epoch, &solution)?;
 				if let Some(ref previous_solution) = epoch.best_submission {
 					ensure!(
-						&new_solution > previous_solution,
+						&new_solution >= previous_solution,
```
Comment (offerijns) [inline pallets/pools/src/lib.rs:1030]: I wonder about one thing: if there are multiple solutions with the same score for an epoch, this makes it possible to keep overriding the best solution. I'm not sure if that's an issue. I don't think so, but maybe worth a little bit of thought.

A potential solution would be to only allow `>=` once.

File: pallets/pools/src/tests.rs [hunk 2]
```diff
@@ -803,24 +803,21 @@ fn submission_period() {
...
-					},
-					TrancheSolution {
-						invest_fulfillment: Perquintill::one(),
-						redeem_fulfillment: Perquintill::one(),
-					}
-				]
-			),
-			Error::<Test>::NotNewBestSubmission
-		);
+		// Can submit the same solution twice
```
Comment (offerijns) [inline pallets/pools/src/tests.rs:806]: Maybe you could add a test case specifically for the deadlock scenario? Basically copying https://github.com/centrifuge/tinlake/blob/develop/src/lender/test/coordinator-submitEpoch.sol#L250 and/or https://github.com/centrifuge/tinlake/blob/develop/src/lender/test/coordinator-submitEpoch.sol#L290

File: pallets/pools/src/lib.rs (hunk 1 above)
Comment (offerijns) [reply pallets/pools/src/lib.rs:1030]: Never mind, I don't think this is an issue.

**Review by offerijns — APPROVED** [review]

File: pallets/pools/src/tests.rs (hunk 2 above)
Comment (mustermeiszer) [reply pallets/pools/src/tests.rs:806]: Not sure I am getting, that this is the deadlock scenario. With the current `>=` there can not be a deadlock as assured by the test above. But I can add a test that allows to submit a zero-solution as the first one, to trigger the `ChallangePeriod`. WDYT?

**Comment by mustermeiszer** [comment]
@offerijns and @branan  I changed the `execute_epoch` error resolution slightly. Are you fine with this? Then you can hit the green button.

**Comment by offerijns** [comment]
> @offerijns and @branan I changed the `execute_epoch` error resolution slightly. Are you fine with this? Then you can hit the green button.

Yeah good idea, this is more explicit 👍

File: pallets/pools/src/tests.rs (hunk 2 above)
Comment (mustermeiszer) [reply pallets/pools/src/tests.rs:806]: Ah, yes. That makes total sense. I will append this scenario to this pr.
